### DISCUSSION TITLE
Support newer CMake versions in Spack minor releases?

### DISCUSSION THREAD

**Discussion by bartlettroscoe** [spack/spack#29783] [General]
I see some documentation about the Spack release process involving branches `releases/vX.Y` at:

* https://github.com/spack/spack#releases
* https://spack.readthedocs.io/en/latest/developer_guide.html#releases

but where is there a discussion of the rest of the workflow like the usage of `backports/vX.Y.Z` branches?

From the looks of it with  `backports/v0.17.2`, it seems like these `backports/vY.Y.Z` branches contain cherry-picks of commits from `develop` that you want to put on the `releases/vY.Y` branch targeted to the `vX.Y.Z` release.  Is that correct?  Is that how one implements:

> Minor releases are back-ported from develop onto release branches. This is typically done by cherry-picking bugfix commits off of `develop`.

?

I am asking because it would be great if the next Spack release 0.17.2 could support CMake 3.22.  But allowing a newer version of CMake would violate the policy:

> To avoid version churn for users of a release series, minor releases should not make changes that would change the concretization of packages. They should generally only contain fixes to the Spack core. However, sometimes priorities are such that new functionality needs to be added to a minor release.

Is it not possible to add new versions of Spack packages like CMake without impacting concretization?  Could one not add new versions of a package that can only be selected by explicitly listing their version like with `cmake@3.22.3`?  That is all we need here.

But in fact, it would be great if the Spack cmake/packages.py file would allow the install of arbitrary versions of CMake pulled from tags off the repo https://github.com/kitware/cmake/.  Same goes for Ninja.

It should be that Spack should not get in the way of allowing users to install the most current versions of simple software like CMake and Ninja just because of the Spack release process cycle.  For these simple packages, Spack should be able to operate as a convenient tool to download, configure, build, and install the software for arbitrary versions of that software (pulled form tags in their native git repos).

NOTE: It looks like from https://github.com/spack/spack/projects/45 it looks like it might be a long time before Spack 0.18.0 gets released which I would assume `releases/v0.18` would be branched off of `develop` at some point.

**Comment by haampie** [comment]
> Is it not possible to add new versions of Spack packages like CMake without impacting concretization?

We just backport commits, and then the answer is no, because default concretization picks up the latest cmake versions.

> But in fact, it would be great if the Spack cmake/packages.py file would allow the install of arbitrary versions of CMake pulled from tags off the repo

Did you even try? Almost all versions are there. Currently 3.23.0 is missing because it was released less than a day ago. But even this one can be installed with `spack install cmake@3.23.0`, spack will simply warn that it has no checksum.

**Reply by bartlettroscoe** [reply]
> Did you even try? Almost all versions are there. 

@haampie, you are referring to the 'develop' branch.  The problem is that there will not be an official release of Spack that supports CMake 3.22 for a long time.  And there are DevOps teams that will only use official releases of Spack so there is no hope of getting CMake 3.22+ for a long time from Spack.

What is needed for Spack packages like CMake and Ninja to allow the user to pass in an arbitrary tag and if that does not match one of the specifically named versions in the package's package.py file, then it would go out to the official git repos for those packages and grab them from there.   I wrote a tool called [`install-cmake.py`](https://github.com/TriBITSPub/TriBITS/blob/master/tribits/devtools_install/install-cmake.py) a long time ago that does that.  So I never have to touch anything as new versions come out.  You can see the implementation of that at:

https://github.com/TriBITSPub/TriBITS/blob/b74b83b2581ab9acf67149322be72f488baca8e9/tribits/devtools_install/InstallProgramDriver.py#L360-L370

https://github.com/TriBITSPub/TriBITS/blob/b74b83b2581ab9acf67149322be72f488baca8e9/tribits/devtools_install/install-cmake.py#L175-L178

I could put in a PR against Spack 'develop' that implements this for `var/spack/repos/builtin/packages/cmake/package.py` (and `var/spack/repos/builtin/packages/ninja/package.py` for that matter) which would solve this problem once and for all.  That would not impact Spack 0.17.z releases but it would impact Spack 0.18+ releases and would solve the CMake deployment problem with Spack in all future Spack releases.

When is Spack 0.18.0 slated to be released?

**Comment by bartlettroscoe** [comment]
Correct me if I am wrong, but from looking at:

* https://github.com/spack/spack/tags

it seems that major releases of Spack v0.Y.0 have slowed to coming out about once a year as shown by the recent releases:

* **v0.17.0**: 11/05/2021
* **v0.16.0**: 11/18/2020
* **v0.15.0**: 06/30/2020
* **v0.14.0**: 02/23/2020
* **v0.13.0**: 11/05/2020

?

So can we expect Spack 0.18.0 to be released sometime in Nov of 2022?

I am guessing that greater stability requirements for Spack releases that goes with more usage by customers is causing the lengthening of time between major Spack v0.Y.0 releases?

**Comment by bartlettroscoe** [comment]
Replying to comment:

* https://github.com/trilinos/Trilinos/issues/10355#issuecomment-1083106538

from @haampie here (to avoid further subscribing him to that GitHub Issue).

> @bartlettroscoe please stop pinging me all over the place on Github, once is enough. 

@haampie, I apologize.  You commented in that Trilinos GitHub Issue so I assumed you did not mind being mentioned there.  Note that you have to @mention someone on GitHub to ensure they get a notification in case they unsubscribed to the issue (which is impossible for another user to see).  But from now on, I will restrict @mentions to this Discussion issue here.

>  Spack core and the builtin Spack repo are currently tied together. Seems like you want a stable Spack core, but rolling releases for packages. Currently that's simply not supported.

@haampie, but it could be supported as I outlined [above](https://github.com/spack/spack/discussions/29783#discussioncomment-2469712).  Are the Spack maintainers willing to consider and discuss that?

**Comment by bartlettroscoe** [comment]
From:

* https://spack.readthedocs.io/en/v0.17.1/packaging_guide.html

which says:

> Spack also provides untrusted download methods: tarball URLs may be supplied without a checksum, or Git downloads may specify a branch or tag instead of a hash. If the user does not control or trust the source of an untrusted download, it is a security risk. Unless otherwise specified by the user for special cases, Spack should by default use only trusted download methods.

it sounds like Spack should already support the notion of downloading and installing "untrusted" tags from a git repo.

**Comment by bartlettroscoe** [comment]
FYI: I talked with an internal DevOps team yesterday and they mentioned one path forward is to copy the Spack package directory for CMake `spack/var/spack/repos/builtin/packages/cmake/` file (along with its `package.py` file),  put it into a custom Spack "Package Repository", and then make the changes that we need in that copy.  That would override the default  `spack/var/spack/repos/builtin/packages/cmake/` directory and use that for CMake installs.

That is not a great long-term solution but it is something that could allow DevOps teams to install CMake 3.22+ on various platforms using the current Spack v0.17.1 (and other v0.17.z patch releases).

**Comment by bartlettroscoe** [comment]
In https://github.com/trilinos/Trilinos/issues/10355#issuecomment-1084910094 @haampie mentioned that Spack 0.17.1 already supports installing arbitrary versions of CMake using, for example, `spack install  cmake@3.23.0`.   I tried this out on an SNL RHEL7 machine with:

```
$ spack external find
$ spack install --no-checksum cmake@3.23.0~openssl
...
[+] /home/rabartl/Spack.base/spack/opt/spack/linux-rhel7-haswell/gcc-4.8.5/cmake-3.23.0-se23sydc6krjpdoxka2l76n3nmxdr7et
```

It worked like a charm!

I think this issue is resolved!

Just a comment, the `spack external find` command is fantastic.  This really makes Spack very usable.
